### ISSUE TITLE
[EPIC] JIT support for `DataFusion`

### ISSUE THREAD

**Issue by alamb** [apache/arrow-datafusion#2703]
**Summary**
TLDR: The key focus of this work is to speed up fundamentally row oriented operations like hash table lookup or comparisons (e.g. [#2427](https://github.com/apache/arrow-datafusion/issues/2427))

**Background**

DataFusion, like many Arrow systems, is a classic "vectorized computation engine" which works quite well for many common operations. The following paper, gives a good treatment on the various tradeoffs between vectorized and JIT's compilation of query plans: https://db.in.tum.de/~kersten/vectorization_vs_compilation.pdf?lang=de

As mentioned in the paper, there are some fundamentally "row oriented" operations in a database that are not typically amenable to vectorization. The "classics" are: Hash table updates in Joins and Hash Aggregates, as well as comparing tuples in sort.

Another example can be found in [these slides](https://github.com/apache/arrow-datafusion/files/8843957/Expression_evaluation.pdf) from [this presentation](https://docs.google.com/presentation/d/1owNlmpNpC2-eBd-jEYRCt0L8_sW4PWYx70gqnFC0twM/edit#slide=id.gba4e8a221c_0_125)

@yjshen added initial support for JIT'ing in https://github.com/apache/arrow-datafusion/pull/1849 and it currently lives in https://github.com/apache/arrow-datafusion/tree/master/datafusion/jit. He also added partial  support for aggregates in https://github.com/apache/arrow-datafusion/pull/2375 

This ticket aims to be a central location for tracking the status of JIT compiling expressions for anyone who wants to contribute to this effort

**Describe the solution you'd like**
- [ ] 
- [ ] https://github.com/apache/arrow-datafusion/pull/2587
- [ ] https://github.com/apache/arrow-datafusion/issues/2122

**Comment by alamb** [comment]
Actually, this is basically the same as https://github.com/apache/arrow-datafusion/issues/1861 so closing in favor of that ticket

**Comment by leoluan2009** [comment]
Hi @alamb , do we need to support expression JIT for performance like ClickHouse: https://clickhouse.com/blog/clickhouse-just-in-time-compiler-jit

**Comment by faucct** [comment]
Though the paper that you have mentioned admits that JIT-compilation is beneficial for OLTP workloads:

> Besides OLAP performance, other factors also play an important role. Compilation-based engines have advantages in
< OLTP as they can create fast stored procedures

If DataFusion would have JIT, then it could be useful for building Online ML Feature Store engines.

**Comment by alamb** [comment]
> If DataFusion would have JIT, then it could be useful for building Online ML Feature Store engines.

FWIW there is no reason you couldn't JIT compile arbitrary expressions and run them as UDFs. The API is now basically complete: https://docs.rs/datafusion/latest/datafusion/logical_expr/trait.ScalarUDFImpl.html

It would also be possible to do the same with aggregates / windows / etc

**Comment by faucct** [comment]
I think that compiling SQL-expressions to UDFs by hand would kinda kill the whole point of the framework, but it seems like most of the framework would be irrelevant for the in-memory transformation of Online Features, so I guess it would be easier to build the same thing from scratch, though using the same ideas, like for example Arrow format for intermediate data representation.

**Comment by zero-element** [comment]
I agree with @faucct that this approach could be beneficial for building Online ML Feature Store engines. This is because such systems often involve numerous element-wise calculations, which can be memory bound (considering low flops/bytes ratio with result of each operation needs to be written to memory).

JIT is not inherently in conflict with vectorized code. In fact, the paper referenced earlier also mentioned:

> it is sometimes beneficial to fuse the loops of two (or more) VectorWise-style primitives into a single loop—saving materialization step

This concept may be similar to Whole-Stage Code Generation in Apache Spark. However, I am not entirely certain about this similarity without inspecting the actual implementation details.
